core/src/main/clojure/xtdb/indexer/temporal_log_indexer.clj

```diff
@@ -0,0 +1,153 @@
+                           (types/->field "$data" (ArrowType$Union. UnionMode/Dense (int-array (range 3))) true
+                                          (types/col-type->field "put" [:struct {'iid [:fixed-size-binary 16]
+                                                                                 'row-id :i64
+                                                                                 'valid-from nullable-inst-type
+                                                                                 'valid-to nullable-inst-type}])
+                                          (types/col-type->field "delete" [:struct {'iid [:fixed-size-binary 16]
+                                                                                    'valid-from nullable-inst-type
+                                                                                    'valid-to nullable-inst-type}])
+                                          (types/col-type->field "evict" [:struct {'iid [:fixed-size-binary 16]}])))]))
+
+(defn long->byte-hash [^long l]
+  (byte-array 16 (.getBytes (Long/toHexString l))))
+
+(defmethod ig/init-key :xtdb.indexer/temporal-log-indexer [_ {:keys [^BufferAllocator allocator, ^ObjectStore object-store]}]
+  (let [log-root (VectorSchemaRoot/create temporal-log-schema allocator)
+        log-writer (vw/root->writer log-root)
+        transcient-log-root (VectorSchemaRoot/create temporal-log-schema allocator)
+        transient-log-writer (vw/root->writer transcient-log-root)
+
+        tx-id-wtr (.writerForName transient-log-writer "tx-id" :i64)
+        system-time-wtr (.writerForName transient-log-writer "system-time" [:timestamp-tz :micro "UTC"])
+
+        tx-ops-wtr (.writerForName transient-log-writer "tx-ops")
+
+        tx-ops-el-wtr (.listElementWriter tx-ops-wtr)
+
+        put-wtr (.writerForTypeId tx-ops-el-wtr (byte 0))
+        put-iid-wtr (.structKeyWriter put-wtr "iid")
+        put-row-id-wtr (.structKeyWriter put-wtr "row-id")
+        put-vf-wtr (.structKeyWriter put-wtr "valid-from")
+        put-vt-wtr (.structKeyWriter put-wtr "valid-to")
+
+        delete-wtr (.writerForTypeId tx-ops-el-wtr (byte 1))
+        delete-iid-wtr (.structKeyWriter delete-wtr "iid")
+        delete-vf-wtr (.structKeyWriter delete-wtr "valid-from")
+        delete-vt-wtr (.structKeyWriter delete-wtr "valid-to")
+
+        evict-wtr (.writerForTypeId tx-ops-el-wtr (byte 2))
+        evict-iid-wtr (.structKeyWriter evict-wtr "iid")
+
+        block-row-counts (ArrayList.)
+        !block-row-count (AtomicInteger.)]
+
+    (reify ILogIndexer
+      (startTx [_ tx-key]
+        (.writeLong tx-id-wtr (.tx-id tx-key))
+        (vw/write-value! (.system-time tx-key) system-time-wtr)
+
+        (.startList tx-ops-wtr)
+        (reify ILogOpIndexer
+          (logPut [_ iid row-id app-time-start app-time-end]
+            (.startStruct put-wtr)
+            (.writeBytes put-iid-wtr (ByteBuffer/wrap (long->byte-hash iid)))
```

Comment:
I currently still do this hacky long->byte-hash conversion. What should the iid be that comes in here.